src/packages/cli/src/detach.ts

```diff
@@ -219,52 +221,70 @@ export async function getDetachedInstances(): Promise<DetachedInstance[]> {
   const instances: DetachedInstance[] = [];
 
   const loadingInstancesInfos = dirEntries.map(async dirEntry => {
-    const { name: instanceName, ext } = path.parse(dirEntry.name);
+    const filename = dirEntry.name;
+    const { name: instanceName, ext } = path.parse(filename);
 
-    if (dirEntry.isDirectory() || ext !== ".json") {
-      try {
-        rm(path.join(dataPath, dirEntry.name), { recursive: true });
-        console.warn(`Invalid instance file found; ${dirEntry.name} removed.`);
-      } catch {
-        console.warn(
-          `Invalid instance file found; ${dirEntry.name} could not be removed.`
-        );
-      }
+    let failureReason: string;
+
+    if (ext !== ".json") {
+      failureReason = `"${filename}" does not have a .json extension`;
     } else {
-      let shouldRemoveInstance = false;
+      let instance: DetachedInstance;
       try {
         // getDetachedInstanceByName() throws if the instance file is not found or
         // cannot be parsed
-        const instance = await getDetachedInstanceByName(instanceName);
-
+        instance = await getDetachedInstanceByName(instanceName);
+      } catch (err: any) {
+        failureReason = err.message;
+      }
+      if (instance) {
         const matchingProcess = processes.find(p => p.pid === instance.pid);
         if (!matchingProcess) {
-          console.warn(
-            `Process with PID ${instance.pid} could not be found; removing ${instanceName} from recorded instances.`
-          );
-          shouldRemoveInstance = true;
+          failureReason = `Process with PID ${instance.pid} could not be found`;
         } else if (matchingProcess.cmd !== instance.cmd) {
-          // if the cmd does not match the instance, the process has been killed,
-          // and another application has taken the pid
-          console.warn(
-            `Process with PID ${instance.pid} doesn't match ${instanceName}; removing${instanceName} from recorded instances.`
-          );
-          shouldRemoveInstance = true;
+          failureReason = `Process with PID ${instance.pid} does not match ${instanceName}`;
         } else {
           instances.push(instance);
         }
-      } catch {
-        console.warn(
-          `Failed to load instance data; ${instanceName} has been removed.`
-        );
-        shouldRemoveInstance = true;
       }
-      if (shouldRemoveInstance) await removeDetachedInstanceFile(instanceName);
+    }
+
+    if (failureReason !== undefined) {
+      someInstancesFailed = true;
+      const fullPath = path.join(dataPath, filename);
+      let resolution: string;
+      if (dirEntry.isDirectory()) {
+        const reason = `"${filename}" is a directory`;
+        try {
+          await rm(fullPath, { recursive: true });
```

Comment:
Can this be used with the `force: true` option, so there's no need for `unlink` as well? It seems `rm` should work on both files and folders... but oops, is it available in [node 12](https://nodejs.org/dist/latest-v18.x/docs/api/fs.html#fsrmpath-options-callback)?

Reply:
Ah, great catch! This party doesn't have tests yet.

Reply:
I've created https://github.com/trufflesuite/ganache/issues/3965 to track this improvement.

Unfortunately rm wasn't added until 14.14 :(

I've updated the call to rm() to use `rmDir` to be nice to node 12 (even though we handled that case reasonably well).

Reply:
I'm working on integration tests for this here: #3735 - they work, and they have reasonable coverage, but they are not especially pretty yet.